File: src/enigma.cpp
#include<algorithm>
#include<iostream>
#include<cassert>
#include<string>
#include<vector>
#include<array>

using std::array, std::string, std::vector;

class Enigma
{
    const string wA = ".ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
    const string wB = ".A1YZD8BX2VLF7IOK0TW3EUMC9SRP4GJN6H5Q";
    const string wC = ".YXZACB1ED2GF3IH4KJ5ML6ON7QP8SR9UT0WV";
    enum wheel { ABC, ACB, BAC, BCA, CAB, CBA };
    std::unordered_map<string, array<array<int,3>, CBA+1>> lookup_table;
    const int teeth = wA.length();
    string const *wl, *wm, *wr; 

    int modulo(int a) { return (teeth + (a % teeth)) % teeth; }
    
    auto& load_state(string& key, int wheel_pos)
    {
        assert(lookup_table.find(key) != lookup_table.end());
        assert(put_wheel(wheel_pos));
        return lookup_table[key][wheel_pos];
    }
    
    bool put_wheel(int comb)
    {
        if(comb<(int)ABC || comb>(int)CBA){
            std::fprintf(stderr, "Error: wheel config\n");
            return false;
        }
        switch (comb)
        {
            case ABC:
                wl = &wA; wm = &wB; wr = &wC; break;
            case ACB:
                wl = &wA; wm = &wC; wr = &wB; break;
            case BAC:
                wl = &wB; wm = &wA; wr = &wC; break;
            case BCA:
                wl = &wB; wm = &wC; wr = &wA; break;
            case CAB:
                wl = &wC; wm = &wA; wr = &wB; break;
            case CBA:
                wl = &wC; wm = &wB; wr = &wA; break;
            default:
                assert(false);
        }
        return true;
    }

    void rotate(int w, array<int,3>& start, int target) // bind physics
    {
        int sign = (w%2 == 0) ? 1 : -1;
        int dist = modulo(target - start[w]);
        for(auto& c : start){
            c = modulo(c + sign * dist);
            sign *= -1;
        }
    }

    bool validate(string& key)
    {
        bool valid = (key.length()==3);
        valid &= (lookup_table.find(key) != lookup_table.end());
        if(!valid) std::fprintf(stderr, "Error: key\n");
        return valid;
    }

    bool verify(string& msg)
    {
        bool valid = true;
        for(auto c : msg)
            valid &= (wA.find_first_of(c) != string::npos);
        if(!valid) std::fprintf(stderr, "Error: message\n");
        return valid;
    }

public:
    Enigma()
    {
        assert(teeth == wB.length());
        assert(teeth == wC.length());
        lookup_table.reserve(teeth * teeth * teeth);
        for(int comb=ABC; comb<=CBA; comb++){
            assert(put_wheel(comb));
            for(int i=0; i<teeth; i++){
                for(int j=0; j<teeth; j++){
                    for(int k=0; k<teeth; k++){
                        string key({wl->at(i), wm->at(j), wr->at(k)});
                        array<int, 3> state({i, j, k});
                        lookup_table[key].at(comb) = state;
                    }
                }
            }
        }
        wl = nullptr; wm = nullptr; wr = nullptr;
    }

    string encrypt(string key, int wheel_pos, string msg)
    {
        if(!put_wheel(wheel_pos) || !validate(key) || !verify(msg))
            return "";
        auto state = load_state(key, wheel_pos);
        string const* readers[2] = {wl, wm};
        string cypher;
        for(int p=0; p<msg.length(); p++){
            rotate(2, state, wr->find_first_of(msg.at(p)));
            cypher += readers[p%2]->at(state[p%2]);
        }
        return cypher;
    }

    string decrypt(string key, int wheel_pos, string msg)
    {
        if(!put_wheel(wheel_pos) || !validate(key) || !verify(msg))
            return "";
        auto state = load_state(key, wheel_pos);
        string const* readers[2] = {wl, wm};
        string plaintext;
        for(int p=0; p<msg.length(); p++){
            rotate(p%2, state, readers[p%2]->find_first_of(msg.at(p)));
            plaintext += wr->at(state[2]);
        }
        return plaintext;
    }
};

File: script/enigma.py
from itertools import product

class Enigma:
    def __init__(self):
        self.wA = ".ABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890"
        self.wB = ".A1YZD8BX2VLF7IOK0TW3EUMC9SRP4GJN6H5Q"
        self.wC = ".YXZACB1ED2GF3IH4KJ5ML6ON7QP8SR9UT0WV"
        self.teeth = len(self.wA)
        self.wheels = [ None, None, None ]
        self.wheel_order = {"ABC": 1, "ACB": 2, "BAC": 3, "BCA": 4, "CAB": 5, "CBA": 6}
        self.lookup_table = {f"{self.wA[i]}{self.wB[j]}{self.wC[k]}":
            [] for i, j, k in product(range(self.teeth), repeat=3)}
        self.init_lookup_table()


    def help(self):
        print("""
usage: 
a: cyphertext (decrypt 6 * 37^2 times)
b: wheel cyphertext (decrypt 37^2 times)
c: key cyphertext (decrypt 6 times)
d: decrypt, wheel key cyphertext
e: encrypt, wheel key plaintext
h: print this help message
q: quit
        """)

    def all(self, args):
        usage = "Usage: a <cyphertext>"
        if len(args) < 1:
            print("Invalid number of arguments.\n" + usage)
            return
        result = [self._shared(" ".join(args),
            self.parse_order_key(order+1, key)) for key in
            list(self.lookup_table.keys())[:self.teeth * self.teeth]
            for order in range(6)]
        result = list(set(result))
        if len((result)) != self.teeth * self.teeth * 6:
            print(f"{len(result)} != {self.teeth * self.teeth * 6}.")
        print(result)

    def hint_wheel(self, args):
        usage = "Usage: b <wheel_order> <cyphertext>"
        if len(args) < 2:
            print("Invalid number of arguments.\n" + usage)
            return
        order = self.validate_wheel(args[0])
        if order is None:
            return
        result = [self._shared(" ".join(args[1:]),
            self.parse_order_key(order+1, key)) for key in
            list(self.lookup_table.keys())[:self.teeth*self.teeth]]
        result = list(set(result))
        if len((result)) != self.teeth * self.teeth:
            print(f"{len(result)} != {self.teeth * self.teeth}.")
        print(result)

    def hint_key(self, args):
        usage = "Usage: c <key> <cyphertext>"
        if len(args) < 2:
            print("Invalid number of arguments.\n" + usage)
            return
        key = self.validate_key(args[0])
        if key is None:
            return
        print([self._shared(" ".join(args[1:]),
                self.parse_order_key(wheel_order+1, key))
                for wheel_order in range(6)])

    def decrypt(self, args):
        usage = "Usage: d <wheel_order> <key> <cyphertext>"
        if len(args) < 3:
            print("Invalid number of arguments.\n" + usage)
            return
        state = self.parse_order_key(args[0], args[1])
        if state is None:
            return
        print(self._shared(" ".join(args[2:]), state))

    def encrypt(self, args):
        usage = "Usage: e <wheel_order> <key> <plaintext>"
        if len(args) < 3:
            print("Invalid number of arguments.\n" + usage)
            return
        state = self.parse_order_key(args[0], args[1])
        if state is None:
            return
        print(self._shared(" ".join(args[2:]), state, decrypt=False))

    def _shared(self, cyphertext, state, decrypt=True):
        # Record the positions of characters that are not in the wheel
        positions = [ i for i, char in enumerate(cyphertext)
                        if char.upper() not in self.wA ]
        # Remove characters that not in the wheel from the message
        msg = [char.upper() for char in cyphertext if char.upper() in self.wA]

        ### simulate rotation of wheels ###
        if decrypt:
            for i in range(len(msg)):
                self.rotate(i%2, state, self.wheels[i%2].index(msg[i]))
                msg[i] = self.wheels[2][state[2]]
        else:# encrypt
            for i in range(len(msg)):
                self.rotate(2, state, self.wheels[2].index(msg[i]))
                msg[i] = self.wheels[i%2][state[i%2]]

        # Insert characters not in the wheel back into the message
        [ msg.insert(_, cyphertext[_]) for _ in positions ]
        return "".join(msg)

    def parse_order_key(self, order, key):
        order = self.validate_wheel(order)
        if order is None:
            return None

        key = self.validate_key(key)
        if key is None:
            return None

        self.put_wheel(order)
        return list(self.lookup_table[key][order])

    def init_lookup_table(self):
        for comb in range(6):
            self.put_wheel(comb)
            for i, j, k in product(range(self.teeth), repeat=3):
                key = f"{self.wheels[0][i]}{self.wheels[1][j]}{self.wheels[2][k]}"
                self.lookup_table[key].append((i, j, k))

    def validate_wheel(self, order):
        if isinstance(order, int) and 1 <= order <= 6:
            return order - 1
        elif order.isdigit() and 1 <= int(order) <= 6:
            return int(order) - 1
        elif order.upper() in self.wheel_order:
            return self.wheel_order[order.upper()] - 1
        else:
            print(f"Invalid wheel order {order} (case insensitive).\n"+\
            "Must be one of ABC, ACB, BAC, BCA, CAB, CBA or an integer from 1 to 6.")
            return

    def validate_key(self, key):
        if key.upper() not in self.lookup_table:
            print(f"Invalid key {key}.\n"+\
            "Must be a 3-character key from the wheel characters.")
            return
        return key.upper()

    def put_wheel(self, comb):
        if comb == 0:
            self.wheels = [ self.wA, self.wB, self.wC ]
        elif comb == 1:
            self.wheels = [ self.wA, self.wC, self.wB ]
        elif comb == 2:
            self.wheels = [ self.wB, self.wA, self.wC ]
        elif comb == 3:
            self.wheels = [ self.wB, self.wC, self.wA ]
        elif comb == 4:
            self.wheels = [ self.wC, self.wA, self.wB ]
        elif comb == 5:
            self.wheels = [ self.wC, self.wB, self.wA ]
        else:
            raise ValueError("Unknown wheel combination")
        
    def rotate(self, w, start, target):
        sign = 1 if w % 2 == 0 else -1
        dist = (target - start[w]) % self.teeth
        start[0] = (start[0] + sign * dist) % self.teeth
        start[1] = (start[1] - sign * dist) % self.teeth
        start[2] = (start[2] + sign * dist) % self.teeth